satellite/gracefulexit/pathcollector.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gracefulexit
+
+import (
+	"context"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/metainfo"
+)
+
+var _ metainfo.Observer = (*PathCollector)(nil)
+
+// PathCollector uses the metainfo loop to add paths to node reservoirs
+//
+// architecture: Observer
+type PathCollector struct {
+	db        DB
+	nodeIDMap map[storj.NodeID]struct{}
+	buffer    []TransferQueueItem
+	log       *zap.Logger
+	batchSize int
+}
+
+// NewPathCollector instantiates a path collector.
+func NewPathCollector(db DB, nodeIDs storj.NodeIDList, log *zap.Logger, batchSize int) *PathCollector {
+	buffer := make([]TransferQueueItem, 0, batchSize)
+	collector := &PathCollector{
+		db:        db,
+		log:       log,
+		buffer:    buffer,
+		batchSize: batchSize,
+	}
+
+	if len(nodeIDs) > 0 {
+		collector.nodeIDMap = make(map[storj.NodeID]struct{}, len(nodeIDs))
+		for _, nodeID := range nodeIDs {
+			collector.nodeIDMap[nodeID] = struct{}{}
+		}
+	}
+
+	return collector
+}
+
+// Flush persists the current buffer items to the database.
+func (collector *PathCollector) Flush(ctx context.Context) (err error) {
+	return collector.flush(ctx, 1)
+}
+
+// RemoteSegment takes a remote segment found in metainfo and creates a graceful exit transfer queue item if it doesn't exist already
+func (collector *PathCollector) RemoteSegment(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) (err error) {
+	if len(collector.nodeIDMap) == 0 {
+		return nil
+	}
+
+	numPieces := int32(len(pointer.GetRemote().GetRemotePieces()))
+	for _, piece := range pointer.GetRemote().GetRemotePieces() {
+		_, ok := collector.nodeIDMap[piece.NodeId]
+		if ok {
+			item := TransferQueueItem{
+				NodeID:          piece.NodeId,
+				Path:            []byte(path.Raw),
+				PieceNum:        piece.PieceNum,
+				DurabilityRatio: float64(numPieces / pointer.GetRemote().GetRedundancy().GetSuccessThreshold()),
```

Comment:
Not 100% sure, but I think durability should be `numPieces / totalThreshold` since `numPieces` can sometimes be larger than success threshold.

Reply:
Also, we should be careful using the value we store for `DurabilityRatio` here. That is probably something we should calculate on the fly when we decide whether or not to transfer a piece - we are only running through the metainfo loop one time for each graceful exiting node, meaning the durability of a segment could change in between when we see each piece and when those pieces are actually transferred.

Reply:
updated